src/lib/schema.js

```diff
@@ -0,0 +1,147 @@
+const validateAsNestedSchemaRules: ValidateAsNestedSchemaRules = field => schema =>
+  validateAsSchemaRules(schema[field]);
+
+/**
+ * Enforces a failed validation if the schema is invalid
+ */
+type ProcessSchemaError = SchemaResult => Array<Validator>;
+export const processSchemaError: ProcessSchemaError = schemaResult => [
+  alwaysErr(getErrors(mapErrors(err => `Schema error: ${err}`)(schemaResult))),
+];
+
+/**
+ * Returns a result, having validated the supplied ObjectSchema
+ */
+type ValidateAsObjectSchema = ObjectSchema => Result;
+export const validateAsObjectSchema: ValidateAsObjectSchema = schema =>
+  allWhileOK([
+    validateIsObject,
+    ...Object.keys(schema).map(validateAsNestedSchemaRules),
+  ])(schema);
+
+/**
+ * Returns a result, having validated the supplied ArraySchema
+ */
+type ValidateAsArraySchema = ArraySchema => Result;
+export const validateAsArraySchema: ValidateAsArraySchema = schema =>
+  allWhileOK([validateIsObject, validateAsSchemaRules])(schema);
+
+/**
+ * Convert an object schema to an array of validators
+ */
+type FromObjectSchema = ObjectSchema => Array<Validator>;
+export const fromObjectSchema: FromObjectSchema = (schema = {}) => {
+  if (!isObject(schema))
+    return processSchemaError(err(['Schemas must be objects']));
+
+  const schemaResult = validateAsObjectSchema(schema);
+  if (isErr(schemaResult)) return processSchemaError(schemaResult);
+
+  const requiredChecks = Object.keys(schema)
+    .filter(k => schema[k].required)
+    .map(validateObjHasKey);
+
+  const typeChecks = Object.keys(schema)
+    .filter(k => schema[k].type)
+    .map(k => validateObjPropHasType(schema[k].type)(k));
+
+  const validatorChecks = Object.keys(schema)
+    .filter(k => schema[k].validator)
+    .map(k => validateObjPropPasses(schema[k].validator)(k));
+
+  return [
+    validateIsObject,
+    all(requiredChecks),
+    all(typeChecks),
+    all(validatorChecks),
+  ];
+};
+
+/**
+ * Convert an object schema to an array of validators and forbid extra fields
+ */
+export const fromObjectSchemaStrict: FromObjectSchema = (schema = {}) => {
+  const vs = fromObjectSchema(schema);
+  return isObject(schema)
+    ? [...vs, validateObjOnlyHasKeys(Object.keys(schema))]
+    : vs;
+};
+
+/**
+ * Convert an array schema to an array of validators
+ */
+type FromArraySchema = ArraySchema => Array<Validator>;
+export const fromArraySchema: FromArraySchema = (schema = {}) => {
+  if (!isObject(schema))
+    return processSchemaError(err(['Schemas must be objects']));
+
+  const schemaResult = validateAsArraySchema(schema);
+  if (isErr(schemaResult)) return processSchemaError(schemaResult);
+
+  return Object.keys(schema)
+    .map(k => {
+      if (k === 'type' && schema[k]) {
+        return [validateArrayItemsHaveType(schema[k])];
+      } else if (k === 'validator' && schema[k]) {
+        return [validateArrayItemsPass(schema[k])];
+      } else return [];
+    })
+    .reduce(flatten, [validateIsArray]);
+};
```

Comment:
`fromObjectSchema` is very clean now – is there any way to rewrite the last part of this in a cleaner way too? I suppose that would just look like

```js
if (schema.type) {...}
if (schema.validator) {...}
```

which may not be much better.